src/main/java/edu/kit/compiler/optimizations/ArithmeticIdentitiesOptimization.java

```diff
@@ -0,0 +1,447 @@
+                exchange(node, node.getRight());
+            } else if (isConstOne(node.getRight())) {
+                // x * 1 --> x
+                exchange(node, node.getLeft());
+
+            } else if (isConstZero(node.getLeft())) {
+                // 0 * x --> 0
+                var zero = node.getRight().getMode().getNull();
+                exchange(node, graph.newConst(zero));
+            } else if (isConstZero(node.getRight())) {
+                // x * 0 --> 0
+                var zero = node.getLeft().getMode().getNull();
+                exchange(node, graph.newConst(zero));
+
+            } else if (isConstNegOne(node.getLeft())) {
+                // -1 * x --> -x
+                var minus = graph.newMinus(node.getBlock(), node.getRight());
+                exchange(node, enqueued(minus));
+            } else if (isConstNegOne(node.getRight())) {
+                // x * -1 --> -x
+                var minus = graph.newMinus(node.getBlock(), node.getLeft());
+                exchange(node, enqueued(minus));
+
+            } else if (isOnlyLeftConst(node.getLeft(), node.getRight())) {
+                // Normalization: c * x --> x * c
+                var mul = graph.newMul(node.getBlock(), node.getRight(), node.getLeft());
+                exchange(node, enqueued(mul));
+
+            } else if (node.getRight().getOpCode() == ir_opcode.iro_Const) {
+                // (x * c) * d --> x * (c * d)
+                foldMultiplication(node);
+            }
+        }
+
+        @Override
+        public void visit(Div node) {
+            if (isConstOne(node.getRight())) {
+                // x / 1 --> x
+                exchangeDivOrMod(node, node.getMem(), node.getLeft());
+            } else if (isConstNegOne(node.getRight())) {
+                // x / -1 --> -x
+                var minus = graph.newMinus(node.getBlock(), node.getLeft());
+                exchangeDivOrMod(node, node.getMem(), enqueued(minus));
+            } else if (isConstZero(node.getLeft())) {
+                // 0 / x --> 0
+                exchangeDivOrMod(node, node.getMem(), node.getLeft());
+            }
+        }
+
+        @Override
+        public void visit(Mod node) {
+            if (isConstAbsOne(node.getRight())) {
+                // x % 1 --> 0
+                // x % -1 --> 0
+                var zero = node.getLeft().getMode().getNull();
+                exchangeDivOrMod(node, node.getMem(), graph.newConst(zero));
+            }
+        }
+
+        @Override
+        public void visit(Conv node) {
+            // ? make this more generic??
+            if (node.getOp().getOpCode() == ir_opcode.iro_Conv) {
+                // remove casts introduced by 64 bit division where possible
+                var op = (Conv) node.getOp();
+                if (node.getMode().equals(Mode.getIs())
+                        && op.getMode().equals(Mode.getLs())
+                        && op.getOp().getMode().equals(Mode.getIs())) {
```

Comment:
Maybe add the reverse case, too? (Ls -> Is -> Ls)

I think this could occur for chained divisions.

Reply:
Removing such a chain of conversions would change the semantics of the program. The value after the two conversions may differ from the original value.

Reply:
Good point